synapse/cryotank.py

```diff
@@ -571,3 +588,609 @@ def init(self, name, conf=None, timeout=None):
+
+    _present_ contains the encoding information about the current indices
+    _deleting_ contains the indices currently being deleted (but aren't done)
+    _progress_ contains how far each index has gotten, how many sucessful props were indexed (which might be different
+    because of missing properties), and how many normalizations failed and is separate because it gets updated a lot
+    more
+    '''
+
+    def __init__(self, dbenv: lmdb.Environment) -> None:
+        '''
+        Creates metadata for all the indices.
+
+        Args:
+            dbenv (lmdb.Environment): the lmdb instance in which to store the metadata.
+
+        Returns:
+            None
+        '''
+
+        self._dbenv = dbenv
+
+        # The table in the database file (N.B. in LMDB speak, this is called a database)
+        self._metatbl = dbenv.open_db(b'meta')
+        is_new_db = False
+        with dbenv.begin(db=self._metatbl, buffers=True) as txn:
+            indices_enc = txn.get(b'indices')
+            progress_enc = txn.get(b'progress')
+        if indices_enc is None or progress_enc is None:
+            if indices_enc is None and progress_enc is None:
+                is_new_db = True
+                indices_enc = s_msgpack.en({'present': {}, 'deleting': []})
+                progress_enc = s_msgpack.en({})
+            else:
+                raise s_exc.CorruptDatabase('missing meta information in index meta')
+
+        indices = s_msgpack.un(indices_enc)
+
+        # The details about what the indices are actually indexing: the datapath and type.
+        self.indices = {k: _MetaEntry(**v) for k, v in indices.get('present', {}).items()}
+        self.deleting = list(indices.get('deleting', ()))
+        # Keeps track (non-persistently) of which indicies have been paused
+        self.asleep = defaultdict(bool)  # type: ignore
+
+        # How far each index has progressed as well as statistics
+        self.progresses = s_msgpack.un(progress_enc)
+        if not all(p in self.indices for p in self.deleting):
+            raise s_exc.CorruptDatabase('index meta table: deleting entry with unrecognized property name')
+        if not all(p in self.indices for p in self.progresses):
+            raise s_exc.CorruptDatabase('index meta table: progress entry with unrecognized property name')
+        if is_new_db:
+            self.persist()
+
+    def persist(self, progressonly=False, txn=None):
+        '''
+        Persists the index info to the database
+
+        Args:
+            progressonly (bool): if True, only persists the progress (i.e. more dynamic) information
+            txn (Optional[lmdb.Transaction]): if not None, will use that transaction to record data.  txn is
+            not committed.
+
+        Returns:
+            None
+        '''
+        d = {'delete': self.deleting,
+             'present': {k: v._asdict() for k, v in self.indices.items()}}
+
+        with contextlib.ExitStack() as stack:
```

Comment:
curious - is this the safe way to enter a lmdb transaction?

Reply:
This is the way to make an optional context. If txn is already provided, we don't need a context, otherwise we do